src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -927,6 +928,56 @@ private static Expression CreateParamCheckingResponseWritingMethodCall(Type retu
         return Expression.Block(localVariables, checkParamAndCallMethod);
     }
 
+    private static void PopulateBuiltInResponseTypeMetadata(Type returnType, IEnumerable<ParameterInfo> parameters, EndpointBuilder builder)
+    {
+        if (returnType.IsByRefLike)
+        {
+            throw GetUnsupportedReturnTypeException(returnType);
+        }
+
+        if (returnType == typeof(Task) || returnType == typeof(ValueTask))
+        {
+            returnType = typeof(void);
+        }
+        else if (AwaitableInfo.IsTypeAwaitable(returnType, out _))
+        {
+            var genericTypeDefinition = returnType.IsGenericType ? returnType.GetGenericTypeDefinition() : null;
+
+            if (genericTypeDefinition == typeof(Task<>) || genericTypeDefinition == typeof(ValueTask<>))
+            {
+                returnType = returnType.GetGenericArguments()[0];
+            }
+            else
+            {
+                throw GetUnsupportedReturnTypeException(returnType);
+            }
+        }
+
+        // Skip void returns and IResults. IResults might implement IEndpointMetadataProvider but otherwise we don't know what it might do.
+        if (returnType == typeof(void) || typeof(IResult).IsAssignableFrom(returnType))
+        {
+            return;
+        }
+
+        // Skip methods that have HttpContext or HttpResponse parameters that might change the status code from 200.
+        foreach (var parameter in parameters)
+        {
+            if (parameter.ParameterType == typeof(HttpContext) || parameter.ParameterType == typeof(HttpResponse))
```

Comment:
REVIEW: Do we like skipping response content type inference for delegates that take an `HttpContext` or `HttpResponse`?

The argument for not inferring is that we don't *really* know the response status code, but it's probably safe to assume the response status code will be 200 most of the time.

Reply:
Not quite following why we'd base that on the parameters accepted (esp. `HttpContext`) as they can often be consumed simply to read from the request rather than manipulate the response, esp. since the method is (probably) returning `IResutl` anyway.

Reply:
This is only for the string and POCO returning cases. RDF itself doesn't infer anything about IResults. It will call the type's PopulateMetadata implementation if it was one regardless of the delegate's parameters.

This is being very conservative saying that if you're both returning a string or a POCO that while also taking an `HttpContext` or `HttpResponse`, one plausible reason to do this is to change the status code. Of course, there are many other reasons to take these parameters, and you are probably sticking with a 200 response either way, but these parameters make that a little less certain.

I can remove this logic though. I called this out for review because it's a pretty non-obvious special case, and I don't like that.

Reply:
Yeah I don't think we want to skip these, otherwise IIUC the following endpoint will not describe itself as returning a 200?

```csharp
app.MapGet("/hello", (HttpContext context) => $"Hello {context.User.Identity.Name}");
```